src/Data/Filters/UserProjectMatch.php

```diff
@@ -0,0 +1,68 @@
+<?php
+/**
+ * This file provides an implementation of the UserProjectMatch filter.
+ *
+ * PHP Version 7
+ *
+ * @category   Data
+ * @package    Main
+ * @subpackage Data
+ * @author     Dave MacFarlane <david.macfarlane2@mcgill.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+namespace LORIS\Data\Filters;
+
+/**
+ * UserProjectMatch filters out data for any resource which is not part of one of the
+ * user's projects. For a DataInstance to be compatible with the UserProjectMatch
+ * filter, it must implement a getProjectIDs or getProjectID method which returns
+ * an integer (or array) of ProjectIDs that the data belongs to. The data will be
+ * filtered out unless the User is a member of at least one project that the resource
+ * DataInstance is a member of.
+ *
+ * @category   Data
+ * @package    Main
+ * @subpackage Data
+ * @author     Dave MacFarlane <david.macfarlane2@mcgill.ca>
+ * @license    http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link       https://www.github.com/aces/Loris/
+ */
+class UserProjectMatch implements \LORIS\Data\Filter
+{
+    /**
+     * Implements the \LORIS\Data\Filter interface
+     *
+     * @param \User                    $user     The user that the data is being
+     *                                           filtered for.
+     * @param \LORIS\Data\DataInstance $resource The data being filtered.
+     *
+     * @return bool true if the user has a project in common with the data
+     */
+    public function filter(\User $user, \Loris\Data\DataInstance $resource) : bool
+    {
+        if (method_exists($resource, 'getProjectIDs')) {
+            // If the Resource belongs to multiple ProjectIDs, the user can
+            // access the data if the user is part of any of those projects.
+            $resourceProjects = $resource->getProjectIDs();
+            foreach ($resourceProjects as $project) {
+                if ($user->hasProject($project)) {
+                    return true;
+                }
+            }
+            return false;
+        } elseif (method_exists($resource, 'getProjectID')) {
+            $resourceProject = $resource->getProjectID();
+            if (!is_null($resourceProject)) {
+                return $user->hasProject($resourceProject);
+            }
```

Comment:
getProjectID() can't return null so I don't think you need this check.

Reply:
All that's been verified by the method_exists check is that there's a function named `getProjectID` on the resource

Reply:
I mean the code hasn't verified anything other than a method exists with that name and the signature can be anything, so the null check is required.

To know anything for sure about the signature, you'd need an interface and an `instanceof` check instead of a method_exists check.